osu.Game/Beatmaps/BeatmapManager.cs

```diff
@@ -89,33 +89,46 @@ public BeatmapManager(Storage storage, IDatabaseContextFactory contextFactory, R
             this.audioManager = audioManager;
         }
 
-        protected override void Populate(BeatmapSetInfo model, ArchiveReader archive)
+        protected override void Populate(BeatmapSetInfo beatmapSet, ArchiveReader archive)
         {
-            model.Beatmaps = createBeatmapDifficulties(archive);
+            if (archive != null)
+                beatmapSet.Beatmaps = createBeatmapDifficulties(archive);
 
-            foreach (BeatmapInfo b in model.Beatmaps)
+            foreach (BeatmapInfo b in beatmapSet.Beatmaps)
             {
                 // remove metadata from difficulties where it matches the set
-                if (model.Metadata.Equals(b.Metadata))
+                if (beatmapSet.Metadata.Equals(b.Metadata))
                     b.Metadata = null;
 
-                // by setting the model here, we can update the noline set id below.
-                b.BeatmapSet = model;
-
-                fetchAndPopulateOnlineIDs(b, model.Beatmaps);
+                b.BeatmapSet = beatmapSet;
             }
 
             // check if a set already exists with the same online id, delete if it does.
-            if (model.OnlineBeatmapSetID != null)
+            if (beatmapSet.OnlineBeatmapSetID != null)
             {
-                var existingOnlineId = beatmaps.ConsumableItems.FirstOrDefault(b => b.OnlineBeatmapSetID == model.OnlineBeatmapSetID);
+                var existingOnlineId = beatmaps.ConsumableItems.FirstOrDefault(b => b.OnlineBeatmapSetID == beatmapSet.OnlineBeatmapSetID);
                 if (existingOnlineId != null)
                 {
                     Delete(existingOnlineId);
                     beatmaps.PurgeDeletable(s => s.ID == existingOnlineId.ID);
-                    Logger.Log($"Found existing beatmap set with same OnlineBeatmapSetID ({model.OnlineBeatmapSetID}). It has been purged.", LoggingTarget.Database);
+                    Logger.Log($"Found existing beatmap set with same OnlineBeatmapSetID ({beatmapSet.OnlineBeatmapSetID}). It has been purged.", LoggingTarget.Database);
                 }
             }
+
+            validateOnlineIds(beatmapSet.Beatmaps);
+
+            foreach (BeatmapInfo b in beatmapSet.Beatmaps)
+                fetchAndPopulateOnlineIDs(b, beatmapSet.Beatmaps);
+        }
+
+        private void validateOnlineIds(List<BeatmapInfo> beatmaps)
+        {
+            var beatmapIds = beatmaps.Where(b => b.OnlineBeatmapID.HasValue).Select(b => b.OnlineBeatmapID);
```

Comment:
Add `ToList()` to this. Don't want to evaluate multiple times in the database query below.